File: packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/classes/options/GetCollectionGroupOptions.java
package io.capawesome.capacitorjs.plugins.firebase.firestore.classes.options;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;

import io.capawesome.capacitorjs.plugins.firebase.firestore.classes.constraints.QueryCompositeFilterConstraint;
import io.capawesome.capacitorjs.plugins.firebase.firestore.classes.constraints.QueryEndAtConstraint;
import io.capawesome.capacitorjs.plugins.firebase.firestore.classes.constraints.QueryLimitConstraint;
import io.capawesome.capacitorjs.plugins.firebase.firestore.classes.constraints.QueryOrderByConstraint;
import io.capawesome.capacitorjs.plugins.firebase.firestore.classes.constraints.QueryStartAtConstraint;
import io.capawesome.capacitorjs.plugins.firebase.firestore.interfaces.QueryNonFilterConstraint;
import org.json.JSONException;

public class GetCollectionGroupOptions {

    @NonNull
    private String reference;

    @Nullable
    private QueryCompositeFilterConstraint compositeFilter;

    @NonNull
    private QueryNonFilterConstraint[] queryConstraints;

    public GetCollectionGroupOptions(String reference, @Nullable JSObject compositeFilter, @Nullable JSArray queryConstraints)
        throws JSONException {
        this.reference = reference;
        if (compositeFilter == null) {
            this.compositeFilter = null;
        } else {
            this.compositeFilter = new QueryCompositeFilterConstraint(compositeFilter);
        }
        if (queryConstraints == null) {
            this.queryConstraints = new QueryNonFilterConstraint[0];
        } else {
            this.queryConstraints = new QueryNonFilterConstraint[queryConstraints.length()];
            for (int i = 0; i < queryConstraints.length(); i++) {
                JSObject queryConstraint = JSObject.fromJSONObject(queryConstraints.getJSONObject(i));
                String queryConstraintType = queryConstraint.getString("type");
                switch (queryConstraintType) {
                    case "orderBy":
                        this.queryConstraints[i] = new QueryOrderByConstraint(queryConstraint);
                        break;
                    case "limit":
                        this.queryConstraints[i] = new QueryLimitConstraint(queryConstraint);
                        break;
                    case "startAt":
                    case "startAfter":
                        this.queryConstraints[i] = new QueryStartAtConstraint(queryConstraint);
                        break;
                    case "endAt":
                    case "endBefore":
                        this.queryConstraints[i] = new QueryEndAtConstraint(queryConstraint);
                        break;
                }
            }
        }
    }

    @NonNull
    public String getReference() {
        return reference;
    }

    @Nullable
    public QueryCompositeFilterConstraint getCompositeFilter() {
        return compositeFilter;
    }

    @NonNull
    public QueryNonFilterConstraint[] getQueryConstraints() {
        return queryConstraints;
    }
}

File: packages/firestore/ios/Plugin/Classes/Options/GetCollectionGroupOptions.swift
import Foundation
import Capacitor

@objc public class GetCollectionGroupOptions: NSObject {
    private var reference: String
    private var compositeFilter: QueryCompositeFilterConstraint?
    private var queryConstraints: [QueryNonFilterConstraint]
    
    init(reference: String, compositeFilter: JSObject?, queryConstraints: Array<JSObject>?) {
        self.reference = reference
        if let compositeFilter = compositeFilter {
            self.compositeFilter = QueryCompositeFilterConstraint(compositeFilter)
        } else {
            self.compositeFilter = nil
        }
        if let queryConstraints = queryConstraints {
            self.queryConstraints = []
            for queryConstraint in queryConstraints {
                let queryConstraintType = queryConstraint["type"] as! String
                switch queryConstraintType {
                case "orderBy":
                    let queryOrderByConstraint = QueryOrderByConstraint(queryConstraint)
                    self.queryConstraints.append(queryOrderByConstraint)
                    break
                case "limit":
                    let queryLimitConstraint = QueryLimitConstraint(queryConstraint)
                    self.queryConstraints.append(queryLimitConstraint)
                    break
                case "startAt", "startAfter":
                    let queryStartAtConstraint = QueryStartAtConstraint(queryConstraint)
                    self.queryConstraints.append(queryStartAtConstraint)
                    break
                case "endAt", "endBefore":
                    let queryEndAtConstraint = QueryEndAtConstraint(queryConstraint)
                    self.queryConstraints.append(queryEndAtConstraint)
                    break
                default:
                    break;
                }
            }
        } else {
            self.queryConstraints = []
        }
    }
    
    func getReference() -> String {
        return reference
    }
    
    func getCompositeFilter() -> QueryCompositeFilterConstraint? {
        return compositeFilter
    }
    
    func getQueryConstraints() -> [QueryNonFilterConstraint] {
        return queryConstraints
    }
}
